### PR TITLE
Harden gather-drop

### PR DESCRIPTION
**Pull request by mmitche**
File.Delete will sometimes fail when pushing a drop to a UNC share. Add a retry for this
Fix a bug where if the first target file location in the list is missing, we'd not copy it on subsequent runs.

### CODE REVIEW

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/GatherDropOperation.cs [hunk 1]
```diff
@@ -1653,5 +1644,24 @@ private static bool IsAzureDevOpsArtifactsUrl(string location)
...
+            await ExponentialRetry.RetryAsync(
+                () =>
+                {
+                    if (File.Exists(filePath))
+                    {
+                        File.Delete(filePath);
+                    }
+                },
+                ex => Console.WriteLine($"Failed to delete {filePath}: {ex.Message}"),
+                ex => ex is UnauthorizedAccessException);
```
Comment (JohnTortugo) [inline src/Microsoft.DotNet.Darc/src/Darc/Operations/GatherDropOperation.cs]: We get `Unauthorized Access` and then it works on retry?

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/GatherDropOperation.cs [hunk 2]
```diff
@@ -1560,10 +1557,7 @@ private static bool IsAzureDevOpsArtifactsUrl(string location)
             // Use a temporary in progress file name so we don't end up with corrupted
             // half downloaded files. Use the first location as the
             string temporaryFileName = $"{targetFiles.First()}.inProgress";
-            if (File.Exists(temporaryFileName))
-            {
-                File.Delete(temporaryFileName);
-            }
+            await DeleteFileWithRetryAsync(temporaryFileName);
```
Comment (JohnTortugo) [inline src/Microsoft.DotNet.Darc/src/Darc/Operations/GatherDropOperation.cs]: I believe you can put some `.ConfigureAwait(false)` in these calls.

Comment (JohnTortugo) [reply src/Microsoft.DotNet.Darc/src/Darc/Operations/GatherDropOperation.cs]: Do we need a try-catch for the case where all retries failed?

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/GatherDropOperation.cs [hunk 3]
```diff
@@ -1616,7 +1610,7 @@ private static bool IsAzureDevOpsArtifactsUrl(string location)
                     // Rename file to the target file name.
                     File.Copy(temporaryFileName, targetFile);
                 }
-                File.Delete(temporaryFileName);
+                await DeleteFileWithRetryAsync(temporaryFileName);
```
Comment (JohnTortugo) [inline src/Microsoft.DotNet.Darc/src/Darc/Operations/GatherDropOperation.cs]: Same observations as for the other call.

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/GatherDropOperation.cs (hunk 1 above)
Comment (mmitche) [reply src/Microsoft.DotNet.Darc/src/Darc/Operations/GatherDropOperation.cs]: Yep.

**Pull request merged**